crates/lair_keystore_api/src/lair_store.rs

```diff
@@ -0,0 +1,420 @@
+#[serde(rename_all = "camelCase")]
+pub struct SeedInfo {
+    /// The ed25519 signature public key derived from this seed.
+    pub ed25519_pub_key: Ed25519PubKey,
+
+    /// The x25519 encryption public key derived from this seed.
+    pub x25519_pub_key: X25519PubKey,
+}
+
+/// The 32 byte blake2b digest of the der encoded tls certificate.
+pub type CertDigest = BinDataSized<32>;
+
+/// Public information associated with a given tls certificate.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct CertInfo {
+    /// The random sni that was generated for this certificate.
+    pub sni: Arc<str>,
+
+    /// The 32 byte blake2b digest of the der encoded tls certificate.
+    pub digest: CertDigest,
+
+    /// The der-encoded tls certificate bytes.
+    pub cert: BinData,
+}
+
+/// The Type and Tag of this lair entry.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(tag = "type", rename_all = "camelCase")]
+#[non_exhaustive]
+pub enum LairEntryInfo {
+    /// This entry is type 'Seed' (see LairEntryInner).
+    Seed {
+        /// user-supplied tag for this seed
+        tag: Arc<str>,
+        /// the seed info associated with this seed
+        seed_info: SeedInfo,
+    },
+
+    /// This entry is type 'DeepLockedSeed' (see LairEntryInner).
+    DeepLockedSeed {
+        /// user-supplied tag for this seed
+        tag: Arc<str>,
+        /// the seed info associated with this seed
+        seed_info: SeedInfo,
+    },
+
+    /// This entry is type 'TlsCert' (see LairEntryInner).
+    WkaTlsCert {
+        /// user-supplied tag for this seed
+        tag: Arc<str>,
+        /// the certificate info
+        cert_info: CertInfo,
+    },
+}
+
+/// The raw lair entry inner types that can be stored.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(tag = "type", rename_all = "camelCase")]
+#[non_exhaustive]
+pub enum LairEntryInner {
+    /// This seed can be
+    /// - derived
+    /// - used for ed25519 signatures
+    /// - used for x25519 encryption
+    /// The secretstream seed uses the base passphrase-derived secret
+    /// for decryption.
+    Seed {
+        /// user-supplied tag for this seed
+        tag: Arc<str>,
+        /// the seed info associated with this seed
+        seed_info: SeedInfo,
+        /// the actual seed, encrypted with context key
+        seed: SecretDataSized<32, 49>,
+    },
+
+    /// As 'Seed' but requires an additional access-time passphrase to use
+    DeepLockedSeed {
+        /// user-supplied tag for this seed
+        tag: Arc<str>,
+        /// the seed info associated with this seed
+        seed_info: SeedInfo,
+        /// salt for argon2id encrypted seed
+        salt: BinDataSized<16>,
+        /// argon2id ops limit used when encrypting this seed
+        ops_limit: u32,
+        /// argon2id mem limit used when encrypting this seed
+        mem_limit: u32,
```

Comment:
what are these two limits for?

Reply:
These are configuration for the argon2 password hashing algorithm, basically how much cpu and how much memory to use. We need to record the values used on the initial creation, so that future unlocks use the same values to arrive at the same hash.